### ISSUE TITLE
One of the request inputs is not valid.

### ISSUE THREAD

**Issue by atrauzzi** [Azure/azure-storage-node#260]
This error...is not helpful.  It's not even traceable from the library outwards.  As soon as I see this, I hit a brick wall and am reduced to guessing at what the problem is.

This error message is so unhelpful, [someone even blogged about it](https://deeperdesign.wordpress.com/2010/03/10/azure-table-storage-what-a-pain-in-the-ass/).

This seems like something that deserves some priority as it's a complete blocker for me right now.  I have no idea what I need to do differently to solve whatever it is I may/may not be doing wrong.

**Comment by stefan-akelius** [comment]
Are you by chance trying to write properties or metadata for a container or blob? The names *and* values have to be ASCII characters. Try to use ASCII-only names and encode all values via url encode or base64.

**Comment by atrauzzi** [comment]
This was for tables, and it looks like a value needed to be stringified.  The library doesn't handle any of that, which is fine, I've fixed that up.

The problem is that I had to spend hours guessing what the problem was.  This waste of time could have been completely avoided if the library and possibly also Azure table storage returned proper feedback in errors.

**Comment by stefan-akelius** [comment]
Yeah, I went through the same time wasting. The error messages and the documentation really have to be updated.

**Comment by XiaoningLiu** [comment]
Hi, we are keeping improve our documentation and error messages. Please reply in this thread for which document or message you feel unclear. It's also welcome anyone who can help sending PRs directly!

**Comment by chugh97** [comment]
I have wasted hours on this now. How do I know which field and cell is causing the issue?

**Comment by XiaoningLiu** [comment]
@chugh97  Can you share a sample code to reproduce this issue?

**Comment by piyush-thakur14** [comment]
Hi,

I am trying to write CSV file to an Azure Blob Storage using Pyspark that I have installed on my VM but while doing so, I am getting this error.

`org.apache.hadoop.fs.azure.AzureException: com.microsoft.azure.storage.StorageException: One of the request inputs is not valid.`

My Pyspark code is as below: -

```
from pyspark import SparkContext
from pyspark.sql import SparkSession
from pyspark.sql import Window
from pyspark.sql.types import *
from pyspark.sql.functions import *
spark = SparkSession.builder.getOrCreate()
storage_account_name = "######"
storage_account_access_key = "####################################3"
spark.conf.set("fs.azure.account.key." + storage_account_name + ".blob.core.windows.net",storage_account_access_key)
spark._jsc.hadoopConfiguration().set("fs.wasbs.impl","org.apache.hadoop.fs.azure.NativeAzureFileSystem")
spark._jsc.hadoopConfiguration().set("fs.azure", "org.apache.hadoop.fs.azure.NativeAzureFileSystem")
spark._jsc.hadoopConfiguration().set("fs.azure.account.key.bdpccadlsdev2.blob.core.windows.net", storage_account_access_key)
df = spark.read.format("csv").option("inferSchema", "true").option("header","true").load("wasbs://<container_name>@<storage_account_name>.blob.core.windows.net/<path_to_csv>")
df.show()
df=df.withColumn("c1",lit("1"))
df.show()
df.coalesce(1).write.mode("overwrite").option("header", "true").format("csv").save("wasbs://<container_name>@<storage_account_name>.blob.core.windows.net/<path_to_write_csv>") 
```

I have used hadoop-azure-2.7.0.jar and azure-storage-2.2.0.jar JARS to read the CSV from my Blob. But I am not able to write back to the blob storage.

Thanks,
Piyush Thakur

**Comment by mirkomiorelli** [comment]
Having the same issue as @piyush-thakur14, only difference is I am writing a parquet. I can read but fail on write...

Any help on this?

Thanks,
Mirko

**Comment by stewartadam** [comment]
@XiaoningLiu any updates on this?

**Comment by XiaoningLiu** [comment]
> Hi,
> 
> I am trying to write CSV file to an Azure Blob Storage using Pyspark that I have installed on my VM but while doing so, I am getting this error.
> 
> `org.apache.hadoop.fs.azure.AzureException: com.microsoft.azure.storage.StorageException: One of the request inputs is not valid.`
> 
> My Pyspark code is as below: -
> 
> ```
> from pyspark import SparkContext
> from pyspark.sql import SparkSession
> from pyspark.sql import Window
> from pyspark.sql.types import *
> from pyspark.sql.functions import *
> spark = SparkSession.builder.getOrCreate()
> storage_account_name = "######"
> storage_account_access_key = "####################################3"
> spark.conf.set("fs.azure.account.key." + storage_account_name + ".blob.core.windows.net",storage_account_access_key)
> spark._jsc.hadoopConfiguration().set("fs.wasbs.impl","org.apache.hadoop.fs.azure.NativeAzureFileSystem")
> spark._jsc.hadoopConfiguration().set("fs.azure", "org.apache.hadoop.fs.azure.NativeAzureFileSystem")
> spark._jsc.hadoopConfiguration().set("fs.azure.account.key.bdpccadlsdev2.blob.core.windows.net", storage_account_access_key)
> df = spark.read.format("csv").option("inferSchema", "true").option("header","true").load("wasbs://<container_name>@<storage_account_name>.blob.core.windows.net/<path_to_csv>")
> df.show()
> df=df.withColumn("c1",lit("1"))
> df.show()
> df.coalesce(1).write.mode("overwrite").option("header", "true").format("csv").save("wasbs://<container_name>@<storage_account_name>.blob.core.windows.net/<path_to_write_csv>") 
> ```
> 
> I have used hadoop-azure-2.7.0.jar and azure-storage-2.2.0.jar JARS to read the CSV from my Blob. But I am not able to write back to the blob storage.
> @XiaoningLiu , Can you or anyone please help me in this?
> Thanks,
> Piyush Thakur

This is not related to Node,.js storage SDK, sorry I cannot help input here.

**Comment by stewartadam** [comment]
@XiaoningLiu can you bring this to the upstream team? Lots of people here are asking for help because the error message is so unspecific.

**Comment by XiaoningLiu** [comment]
For any asks to Azure Storage service, please open and vote in https://feedback.azure.com/forums/217298-storage.

**Comment by stewartadam** [comment]
https://feedback.azure.com/forums/217298-storage/suggestions/3689425-improve-exception-messages

https://feedback.azure.com/forums/217298-storage/suggestions/6517318-let-s-sunset-one-of-the-request-inputs-is-not-val

**Comment by Mangeshshende** [comment]
@pt19034 @mirkomiorelli - did u guys find some solution to above error of writing df?

**Comment by valdisiljuconoks** [comment]
funny enough, we had a somewhat similar issue (but only when deployed to Azure, locally everything was working as expected). turned out - we accessed storage from the API call, our own authorization token was passed in request headers in "Authorization" field. when we switched to ordinary querystring parameter - everything seems to be working fine again. I suspect it might be that request headers are passed further to any subsequent HTTP calls, but we haven't captured and analyzed network traffic to figure this out. it's just our assumption.
